README.md
graw


CHANGELOG.org
#+TITLE: GRAW - Changelog
#+DESCRIPTION: All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) - though syntax has been adapted to Org-mode syntax,
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) - well trying actually.

Futhermore, take note that, in general, if a change has no impact on "users", it is leaved out of this file.
For more internal notes, it may be revelant to check the =LOG= files instead (inside =log= repository).

** [Unreleased]

- 2.5.Z: Update data location / backend.

** [2.4.0] - 2021-01-16

*** Added

- Add participants group / team to the same region.

** [2.3.0] - 2021-01-12

*** Added

- Add multiple (but distincts - no team) participants to the same region.

** [2.2.0] - 2020-12-24

*** Changed

- Drawing process to better actual location rendering.

** 2.2.0 Add multiple participants for one region [YANKED]

GPS traces drawing must be completely reviewed before any other changes.

** [2.1.0] - 2020-12-17

*** Changed

- Distance unit from step to meter.

** [1.3.1] - 2020-12-16

*** Added

- This CHANGELOG to hopefully serve as a good evolution tracker.


utl/test-garminexport.py
#! /usr/bin/env python

import argparse
import getpass
import json
import logging

from garminexport.garminclient import GarminClient

logging.basicConfig(level=logging.INFO, format="%(asctime)-15s [%(levelname)s] %(message)s")
log = logging.getLogger(__name__)

if __name__ == "__main__":

    parser = argparse.ArgumentParser(
        description="Export all Garmin Connect activities")
    # positional args
    parser.add_argument(
        "username", metavar="<username>", type=str, help="Account user name.")
    # optional args
    parser.add_argument(
        "--password", type=str, help="Account password.")

    args = parser.parse_args()
    print(args)

    if not args.password:
        args.password = getpass.getpass("Enter password: ")

    try:
        with GarminClient(args.username, args.password) as client:
            log.info("activities:")
            activity_ids = client.list_activities()
            log.info("num ids: %d", len(activity_ids))
            log.info(activity_ids)

            latest_activity, latest_activity_start = activity_ids[0]
            activity = client.get_activity_summary(latest_activity)
            #log.info("activity id: %s", activity["activity"]["activityId"])
            #log.info("activity name: '%s'", activity["activity"]["activityName"])
            #log.info("activity description: '%s'", activity["activity"]["activityDescription"])
            #log.info(json.dumps(client.get_activity_details(latest_activity), indent=4))
            #log.info(client.get_activity_gpx(latest_activity))
    except Exception as e:
        log.error("failed with exception: %s", e)
    finally:
        log.info("done")

utl/test-garminconnect.py
#!/usr/bin/env python3

from garminconnect import (
    Garmin,
    GarminConnectConnectionError,
    GarminConnectTooManyRequestsError,
    GarminConnectAuthenticationError,
)

from datetime import date
import getpass

"""
Enable debug logging
"""
import logging
logging.basicConfig(level=logging.DEBUG)

today = date.today()

username = input("Username: ")
password = getpass.getpass("Password: ")


"""
Initialize Garmin client with credentials
Only needed when your program is initialized
"""
print("Garmin(email, password)")
print("----------------------------------------------------------------------------------------")
try:    
    client = Garmin(username, password)
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client init: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client init")
    quit()


"""
Login to Garmin Connect portal
Only needed at start of your program
The library will try to relogin when session expires
"""
print("client.login()")
print("----------------------------------------------------------------------------------------")
try:
    client.login()
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client login: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client login")
    quit()


"""
Get full name from profile
"""
print("client.get_full_name()")
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_full_name())
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get full name: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get full name")
    quit()


"""
Get unit system from profile
"""
print("client.get_unit_system()")
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_unit_system())
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get unit system: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get unit system")
    quit()


"""
Get activity data
"""
print("client.get_stats(%s)", today.isoformat())
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_stats(today.isoformat()))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get stats: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get stats")
    quit()


"""
Get steps data
"""
print("client.get_steps_data\(%s\)", today.isoformat())
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_steps_data(today.isoformat()))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get steps data: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get steps data")
    quit()


"""
Get heart rate data
"""
print("client.get_heart_rates(%s)", today.isoformat())
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_heart_rates(today.isoformat()))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get heart rates: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get heart rates")
    quit()


"""
Get body composition data
"""
print("client.get_body_composition(%s)", today.isoformat())
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_body_composition(today.isoformat()))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get body composition: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get body composition")
    quit()


"""
Get stats and body composition data
"""
print("client.get_stats_and_body_composition(%s)", today.isoformat())
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_stats_and_body(today.isoformat()))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get stats and body composition: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get stats and body composition")
    quit()


"""
Get activities data
"""
print("client.get_activities(0,1)")
print("----------------------------------------------------------------------------------------")
try:
    activities = client.get_activities(0,1) # 0=start, 1=limit
    print(activities)
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get activities: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get activities")
    quit()


"""
Download an Activity
"""
try:
    for activity in activities:
        activity_id = activity["activityId"]
        print("client.download_activities(%s)", activity_id)
        print("----------------------------------------------------------------------------------------")

        gpx_data = client.download_activity(activity_id, dl_fmt=client.ActivityDownloadFormat.GPX)
        output_file = f"./{str(activity_id)}.gpx"
        with open(output_file, "wb") as fb:
            fb.write(gpx_data)

        tcx_data = client.download_activity(activity_id, dl_fmt=client.ActivityDownloadFormat.TCX)
        output_file = f"./{str(activity_id)}.tcx"
        with open(output_file, "wb") as fb:
            fb.write(tcx_data)

        zip_data = client.download_activity(activity_id, dl_fmt=client.ActivityDownloadFormat.ORIGINAL)
        output_file = f"./{str(activity_id)}.zip"
        with open(output_file, "wb") as fb:
            fb.write(zip_data)

        csv_data = client.download_activity(activity_id, dl_fmt=client.ActivityDownloadFormat.CSV)
        output_file = f"./{str(activity_id)}.csv"
        with open(output_file, "wb") as fb:
          fb.write(csv_data)
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get activity data: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get activity data")
    quit()


"""
Get sleep data
"""
print("client.get_sleep_data(%s)", today.isoformat())
print("----------------------------------------------------------------------------------------")
try:
    print(client.get_sleep_data(today.isoformat()))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get sleep data: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get sleep data")
    quit()


"""
Get devices
"""
print("client.get_devices()")
print("----------------------------------------------------------------------------------------")
try:
    devices = client.get_devices()
    print(devices)
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get devices: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get devices")
    quit()


"""
Get device settings
"""
try:
    for device in devices:
        device_id = device["deviceId"]
        print("client.get_device_settings(%s)", device_id)
        print("----------------------------------------------------------------------------------------")

        print(client.get_device_settings(device_id))
except (
    GarminConnectConnectionError,
    GarminConnectAuthenticationError,
    GarminConnectTooManyRequestsError,
) as err:
    print("Error occurred during Garmin Connect Client get device settings: %s" % err)
    quit()
except Exception:  # pylint: disable=broad-except
    print("Unknown error occurred during Garmin Connect Client get device settings")
    quit()


index.js
// Strict mode
function strict() {
	'use strict';



	function loadFile(filePath) {
	    return new Promise(resolve => {
	        var result = null;
	        var xmlhttp = new XMLHttpRequest();
    		xmlhttp.open("GET", filePath, false); // 3rd parameter = synchronous
    		xmlhttp.overrideMimeType("text/plain");
    		xmlhttp.send(null);
    		if(xmlhttp.status === 200) {
    			result = xmlhttp.responseText;
    		} else {
    		    console.error("Error loading file '%s'", filePath);
    		}
    		resolve(result);
	    });
	}



	async function loadJsonFile(jsonFilePath) {
        let jsonTxt = await loadFile(jsonFilePath);
        let data = JSON.parse(jsonTxt);
        return data;
	}



	async function displayRegion(region) {
	    let geojsonFilePath = dataFolder + '/' + region.geojsonFile; // @TODO Absolute path...?

	    loadJsonFile(geojsonFilePath).then( function(data) {
            let regionGeojson = data;

            regionGeojsons[region.id] = JSON.parse(JSON.stringify(regionGeojson));

            let regionStyle = {
                "weight": 10
            };

            let regionLayer = L.geoJSON(regionGeojson, {
                style: regionStyle
            });

            regionOverlays[region.id] = regionLayer;
            regionOverlays[region.id].addTo(mymap);
        });
	}



	async function displayPOI(region) {
	    let poiFilePath = dataFolder + '/' + region.poiFile; // @TODO Absolute path...?

	    loadJsonFile(poiFilePath).then( function(data) {
            var poiGeojson = data;

            var geojsonMarkerOptions = {
                icon: null,
                radius: 5,
                fillColor: "blue",
                color: "blue",
                weight: 1,
                opacity: 1,
                fillOpacity: 0.8
            };

            var poiLayer = L.geoJSON(poiGeojson, {
                pointToLayer: function (feature, latlng) {
                    return L.circleMarker(latlng, geojsonMarkerOptions);
                }
            });

            poiOverlays[region.id] = poiLayer;
            poiOverlays[region.id].addTo(mymap);
        });
	}



    async function displayRecord(datetime, data) {
        let record = data;
        let participant = participants.filter((participant) => participant.id === record.id).shift();
        let region = regions[participant.region - 1]; // !!! ID != Index !!!

        /*** TEAM ***/
        if(participant.type === "Team") {
            let progressGeojson = JSON.parse(JSON.stringify(regionGeojsons[region.id])); // "Deep" copy - Seems functional - for coordinates copy at least

            // Record distance is relative to total actual CRAW distance (!= displayed distance)!
            // There is 2 dimensions to consider: Actual/Displayed x Record/Total.

            let distanceActualTotal = region.distanceTrue;

            let distanceActualRecord = 0;

            for(let p of record.participants){
                distanceActualRecord += p.distance;
            }

            // Link between "actual" and "displayed" data
            let rate = distanceActualRecord / distanceActualTotal;

            // Calculate total displayed distance

            let distanceDisplayedTotal = 0;
            let coordinates = progressGeojson.features[0].geometry.coordinates;

            // (1) Create a L.latLng array for easy distance calculus with "distanceTo()"
            let latLngs = [];

            coordinates.forEach(function (coord, i) {
                let latLng = L.latLng(coord[1], coord[0]); // !!! GeoJSON != Leaflet for (lat, lng) variables order !!!
                latLngs.push(latLng);
            });

            // (2) Now actually calculate total distance
            for(let i = 0; i < latLngs.length-1; i++) {
                let distance = latLngs[i].distanceTo(latLngs[i+1]);
            	distanceDisplayedTotal += distance;
            }

            let distanceDisplayedRecord = rate * distanceDisplayedTotal;

            if(distanceDisplayedRecord > distanceDisplayedTotal) { distanceDisplayedRecord = distanceDisplayedTotal; }

            // Calculate closer last waypoint

            let distanceCumul = 0;
            let f = 0; // Index of closer waypoint

            for(let i = 0; i < latLngs.length-1; i++) {
                let distance = latLngs[i].distanceTo(latLngs[i+1]);

            	if(distanceCumul + distance >= distanceDisplayedRecord) { // Progress overcome last waypoint
            	    f = i;
            	    break;
            	} else {
            	    distanceCumul += distance;
            	}
            }

            let lastWaypointIndex = f;
            progressGeojson.features[0].geometry.coordinates.length = lastWaypointIndex + 1; // Probably not the cleanest way to do this

            let distanceRemaining = distanceDisplayedRecord - distanceCumul; // From last covered waypoint

            // Optimal data

            let diffTimeSinceStart = Math.abs(new Date(datetime).getTime() - new Date("2020-11-01").getTime());  // Here is considered the last update as a reference point (and not "today") - also, challenge started on 2020-11-01.
            let diffDaysSinceStart = Math.ceil(diffTimeSinceStart / (1000 * 3600 * 24));
            let distanceActualOptimal = distanceActualTotal / 365 * diffDaysSinceStart;
            let rateOptimal = distanceActualOptimal / distanceActualTotal;

            let msg = '*** Record from participant ' + participant.id + ' for region ' + region.id + ' from ' + participant.type + ' at ' + datetime.toLocaleString() + ' ***\n';
            msg += '+ Participants:\n';
            for(let p of record.participants){
                let pp = participants.filter((participant) => participant.id === p.id).shift();
                msg += '- Participant ' + pp.id + ': "' + pp.name + '"' + '\n';
            };
            msg += '+ Actual distances:\n';
            msg += '- Total: ' + distanceActualTotal + ' m\n';
            msg += '- Record: ' + distanceActualRecord + ' m\n';
            msg += '+ Displayed distances:\n';
            msg += '- Total: ' + distanceDisplayedTotal + ' m\n';
            msg += '- Record: ' + distanceDisplayedRecord + ' m\n';
            msg += '+ Rate: ' + rate * 100 + ' %\n';
            msg += '********************\n';
            msg += '+ Last waypoint: N°' + lastWaypointIndex + ' / ' + progressGeojson.features[0].geometry.coordinates[lastWaypointIndex] + '\n';
            msg += '+ Remaining distance (since last waypoint): ' + distanceRemaining + ' m\n';
            msg += '********************\n';
            msg += '+ Days since start: ' + diffDaysSinceStart + ' d\n';
            msg += '+ Optimal actual distance: ' + distanceActualOptimal + ' m\n';
            msg += '+ Optimal rate: ' + rateOptimal * 100 + ' %\n';
            msg += '************************************************************';
            console.debug(msg);

            /** Display **/

            let progressStyle = {
                "color": "orange",
                "weight": 5,
                "opacity": 0.9
            };

            let progressLayer = L.geoJSON(progressGeojson, {
                style: progressStyle
            }).addTo(mymap);

    		// Add marker at current progress position with information

    		let lastPosition = progressGeojson.features[0].geometry.coordinates.length - 1;

    		let myIcon = L.icon({
    			iconUrl: 'img/pin-icon-runner-team.png',
    			iconSize: [45, 45]
    		});

            let typeImgDiv = "";
            if (participant.type === "Team") {
                typeImgDiv = '<img style="vertical-align: middle;" src="img/pin-icon-runner-team.png" alt="[Team]" title="Pour les participants en équipe sur une même région, les distances sont cumulées." width="24" height="24"></img>';
            }

            let participantsDiv = "";
            for(let p of record.participants){
                let pp = participants.filter((participant) => participant.id === p.id).shift();

                let ppDiv = "";
                if(pp.type === "Garmin") {
                    ppDiv = '<img style="vertical-align: middle;" src="img/logo-garmin-connect.png" alt="[Garmin]" title="Pour les participants Garmin (avec montre), tous les pas réalisés sont pris en compte." width="16" height="16"></img>';
                } else if (pp.type === "Strava") {
                    ppDiv = '<img style="vertical-align: middle;" src="img/logo-strava.png" alt="[Strava]" title="Pour les participants Strava, seules les activités de course ou marche sont pris en compte." width="16" height="16"></img>';
                }

                participantsDiv += "<div> - " + ppDiv + " <span style='vertical-align: middle;'>Participant " + pp.id + "</span></div>";
            };

    		L.marker(latLngs[lastPosition], {
    			icon: myIcon,
    			title: "Region "+participant.id+"\n"+region.title
    		})
    		.bindTooltip(""+participant.id+"", {permanent: true, direction: 'bottom'})
    		.addTo(mymap)
    		.bindPopup(
    		"<div>" + typeImgDiv + "<span style='vertical-align: middle;'><b> Team " + participant.id + " '" + participant.name + "' - Region " + region.id + " (" + region.title + ")" + "</b></span></div>"
    		+"Distance à parcourir : " + (distanceActualTotal/1000).toFixed(2) + " km<br>"
    		+"Distance parcourue : " + (distanceActualRecord/1000).toFixed(2) + " km (" + (rate*100).toFixed(2) + "%)<br>"
    		+"Distance optimale (" + diffDaysSinceStart + "J) : " + (distanceActualOptimal/1000).toFixed(2) + " km (" + (rateOptimal*100).toFixed(2) + "%)<br>"
    		+"Dernière MàJ : " + datetime.toLocaleString() + "<br>"
    		+"Participants :" + "<br>"
    		+ participantsDiv,
    		{
    		    maxWidth : "600"
    		}
    		)
    		;
        }
        /*** SOLO ***/
        else {
            let progressGeojson = JSON.parse(JSON.stringify(regionGeojsons[region.id])); // "Deep" copy - Seems functional - for coordinates copy at least

            // Record distance is relative to total actual CRAW distance (!= displayed distance)!
            // There is 2 dimensions to consider: Actual/Displayed x Record/Total.

            let distanceActualTotal = region.distanceTrue;
            let distanceActualRecord = record.distance;

            // Link between "actual" and "displayed" data
            let rate = distanceActualRecord / distanceActualTotal;

            // Calculate total displayed distance

            let distanceDisplayedTotal = 0;
            let coordinates = progressGeojson.features[0].geometry.coordinates;

            // (1) Create a L.latLng array for easy distance calculus with "distanceTo()"
            let latLngs = [];

            coordinates.forEach(function (coord, i) {
                let latLng = L.latLng(coord[1], coord[0]); // !!! GeoJSON != Leaflet for (lat, lng) variables order !!!
                latLngs.push(latLng);
            });

            // (2) Now actually calculate total distance
            for(let i = 0; i < latLngs.length-1; i++) {
                let distance = latLngs[i].distanceTo(latLngs[i+1]);
            	distanceDisplayedTotal += distance;
            }

            let distanceDisplayedRecord = rate * distanceDisplayedTotal;

            if(distanceDisplayedRecord > distanceDisplayedTotal) { distanceDisplayedRecord = distanceDisplayedTotal; }

            // Calculate closer last waypoint

            let distanceCumul = 0;
            let f = 0; // Index of closer waypoint

            for(let i = 0; i < latLngs.length-1; i++) {
                let distance = latLngs[i].distanceTo(latLngs[i+1]);

            	if(distanceCumul + distance >= distanceDisplayedRecord) { // Progress overcome last waypoint
            	    f = i;
            	    break;
            	} else {
            	    distanceCumul += distance;
            	}
            }

            let lastWaypointIndex = f;
            progressGeojson.features[0].geometry.coordinates.length = lastWaypointIndex + 1; // Probably not the cleanest way to do this

            let distanceRemaining = distanceDisplayedRecord - distanceCumul; // From last covered waypoint

            // Optimal data

            let diffTimeSinceStart = Math.abs(new Date(datetime).getTime() - new Date("2020-11-01").getTime());  // Here is considered the last update as a reference point (and not "today") - also, challenge started on 2020-11-01.
            let diffDaysSinceStart = Math.ceil(diffTimeSinceStart / (1000 * 3600 * 24));
            let distanceActualOptimal = distanceActualTotal / 365 * diffDaysSinceStart;
            let rateOptimal = distanceActualOptimal / distanceActualTotal;

            let msg = '*** Record from participant ' + participant.id + ' for region ' + region.id + ' from ' + participant.type + ' at ' + datetime.toLocaleString() + ' ***\n';
            msg += '+ Actual distances:\n';
            msg += '- Total: ' + distanceActualTotal + ' m\n';
            msg += '- Record: ' + distanceActualRecord + ' m\n';
            msg += '+ Displayed distances:\n';
            msg += '- Total: ' + distanceDisplayedTotal + ' m\n';
            msg += '- Record: ' + distanceDisplayedRecord + ' m\n';
            msg += '+ Rate: ' + rate * 100 + ' %\n';
            msg += '********************\n';
            msg += '+ Last waypoint: N°' + lastWaypointIndex + ' / ' + progressGeojson.features[0].geometry.coordinates[lastWaypointIndex] + '\n';
            msg += '+ Remaining distance (since last waypoint): ' + distanceRemaining + ' m\n';
            msg += '********************\n';
            msg += '+ Days since start: ' + diffDaysSinceStart + ' d\n';
            msg += '+ Optimal actual distance: ' + distanceActualOptimal + ' m\n';
            msg += '+ Optimal rate: ' + rateOptimal * 100 + ' %\n';
            msg += '************************************************************';
            console.debug(msg);

            /** Display **/

            let progressStyle = {
                "color": "orange",
                "weight": 5,
                "opacity": 0.9
            };

            let progressLayer = L.geoJSON(progressGeojson, {
                style: progressStyle
            }).addTo(mymap);

    		// Add marker at current progress position with information

    		let lastPosition = progressGeojson.features[0].geometry.coordinates.length - 1;

    		let myIcon = L.icon({
    			iconUrl: 'img/pin-icon-runner.png',
    			iconSize: [35, 35]
    		});

            let typeImgDiv = "";
            if(participant.type === "Garmin") {
                typeImgDiv = '<img style="vertical-align: middle;" src="img/logo-garmin-connect.png" alt="[Garmin]" title="Pour les participants Garmin (avec montre), tous les pas réalisés sont pris en compte." width="24" height="24"></img>';
            } else if (participant.type === "Strava") {
                typeImgDiv = '<img style="vertical-align: middle;" src="img/logo-strava.png" alt="[Strava]" title="Pour les participants Strava, seules les activités de course ou marche sont pris en compte." width="24" height="24"></img>';
            }

    		L.marker(latLngs[lastPosition], {
    			icon: myIcon,
    			title: "Region "+participant.id+"\n"+region.title
    		})
    		.bindTooltip(""+participant.id+"", {permanent: true, direction: 'bottom'})
    		.addTo(mymap)
    		.bindPopup("<div>" + typeImgDiv + "<span style='vertical-align: middle;'><b> Participant " + participant.id + " - Region " + region.id + " (" + region.title + ")" + "</b></span></div>"
    		+"Distance à parcourir : " + (distanceActualTotal/1000).toFixed(2) + " km<br>"
    		+"Distance parcourue : " + (distanceActualRecord/1000).toFixed(2) + " km (" + (rate*100).toFixed(2) + "%)<br>"
    		+"Distance optimale (" + diffDaysSinceStart + "J) : " + (distanceActualOptimal/1000).toFixed(2) + " km (" + (rateOptimal*100).toFixed(2) + "%)<br>"
    		+"Dernière MàJ : " + datetime.toLocaleString())
    		;
        }
    }

    /**--------------**/
    /** MAIN PROGRAM **/
    /**--------------**/

    /** Init world map **/

	var mymap = L.map('mapid', {
	    zoomSnap: 0.25,
	    worldCopyJump: true
	}).setView([0, 0], 1.75);

	var baseLayer = L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token={accessToken}', {
		maxZoom: 18,
		attribution: 'Map data &copy; <a href="https://www.openstreetmap.org/">OpenStreetMap</a> contributors, ' + '<a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, ' + 'Imagery © <a href="https://www.mapbox.com/">Mapbox</a>',
		id: 'mapbox/streets-v11',
		tileSize: 512,
		zoomOffset: -1,
		accessToken: 'pk.eyJ1IjoibWFwYm94IiwiYSI6ImNpejY4NXVycTA2emYycXBndHRqcmZ3N3gifQ.rJcFIG214AriISLbB6B5aw'
	}).addTo(mymap);

    /* TODO: Layers Leaflet organization
    var layers = {
        "Base": baseLayer
    };

    var regionOverlays = L.layerGroup();
    var poiOverlays = L.layerGroup();

    var overlays = {
        "Regions": regionOverlays,
        "Points Of Interest": poiOverlays
    };

    L.control.layers(layers, overlays).addTo(mymap);
    */

    var regionOverlays = [];
    var poiOverlays = [];
    var progressOverlays = [];
    var regionGeojsons = [];

    /** Load data **/

    // Data files

    const dataFolder = 'data'; // PROD: 'data' ; DEBUG: 'data/mock*'
    const participantsJsonFile = dataFolder + '/' + 'participants.json';
    const regionsJsonFile = dataFolder + '/' +'regions.json';
    const recordsJsonFile = dataFolder + '/' +'records.json';

    console.debug("Data folder: '%s'", dataFolder);
    console.debug("Data files: '%s'", [participantsJsonFile, regionsJsonFile, recordsJsonFile]);

    // Data variables

    var participants = [];
    var regions = [];
    var records = [];

    // Logic variables

    var loadingRegions, loadingParticipants, displayingRegions, loadingRecords; // Promises

    console.debug("[1.A] Loading regions data ('%s')...", regionsJsonFile);
    loadingRegions = loadJsonFile(regionsJsonFile).then( function(data) {
        regions = data;
        console.debug("[1.A] Loaded regions data: %o", regions);
    });

    console.debug("[1.B] Loading participants data ('%s')...", participantsJsonFile);
    loadingParticipants = loadJsonFile(participantsJsonFile).then( function(data) {
        participants = data;
        console.debug("[1.B] Loaded participants data: %o", participants);
    });

    console.debug("Waiting for previous operations [1.*]...");
    Promise.all([loadingRegions, loadingParticipants]).then(([a, b]) => {

        console.debug("[2.A] Displaying regions view...");
        displayingRegions = []
        for(let region of regions) {
            console.debug("[2.A.%d.a] Displaying region %d...", region.id, region.id);
    	    let displayingRegion = displayRegion(region).then( function() {
    	        console.debug("[2.A.%d.a] Displayed region %d", region.id, region.id);
    	    });

    	    displayingRegions.push(displayingRegion);

            /*
            @TODO POI feature
    	    console.debug("[2.A.%d.b] Displaying POI of region %d...", region.id, region.id);
    	    displayPOI(region).then( function() {
    	        console.debug("[2.A.%d.b] Displayed POI of region %d", region.id, region.id);
    	    });
    	    */
        }

        console.debug("[2.B] Loading records data...");
        loadingRecords = loadJsonFile(recordsJsonFile).then( function(data) {
            records = data;
            console.debug("[2.B] Loaded records data: %o", records);
        });

        console.debug("Waiting for previous operations [2.*]...");
        Promise.all([displayingRegions, loadingRecords]).then(([a, b]) => {

            console.debug("[3.A] Finding last record...");
            let lastUpdate = new Date(0);
            let lastUpdateIndex = 0;

            records.forEach(function (record, i) {
                let datetimestr = record.datetime;
                let datetime = new Date(datetimestr + "Z");

                if(datetime > lastUpdate) {
                    lastUpdate = datetime;
                    lastUpdateIndex = i;
                }
            });

            let lastRecord = records[lastUpdateIndex];

            console.debug("[3.A] Found last record: %o", lastRecord);

            console.debug("[3.B] Displaying last record...");

        	lastRecord.data.forEach(function (d, i) {
        	    console.debug("[3.B.%d] Displaying last record data %d...", i, i);
        	    displayRecord(lastRecord.datetime, d).then( function() {
        	        console.debug("[3.B.%d] Displayed last record data %d", i, i);
        	    });
        	});
        });
    });
}
strict();
